lib/system/qnx/irq.c

```diff
@@ -0,0 +1,253 @@
+
+/* Static functions */
+static void metal_qnx_irq_set_enable(struct metal_irq_controller *irq_cntr,
+              int irq, unsigned int state);
+
+/* QNX IRQ controller */
+static METAL_IRQ_CONTROLLER_DECLARE(qnx_irq_cntr,
+            0, MAX_IRQS,
+            NULL,
+            metal_qnx_irq_set_enable, NULL,
+            irqs);
+
+unsigned int metal_irq_save_disable(void)
+{
+	InterruptDisable();
+	return 0;
+}
+
+void metal_irq_restore_enable(unsigned int flags)
+{
+	InterruptEnable();
+	metal_unused(flags);
+}
+
+static void metal_qnx_irq_set_enable(struct metal_irq_controller *irq_cntr,
+              int irq, unsigned int state)
+{
+	if (irq < irq_cntr->irq_base ||
+			irq >= irq_cntr->irq_base + irq_cntr->irq_num) {
+		metal_log(METAL_LOG_ERROR, "%s: invalid irq %d\n",
+			  __func__, irq);
+		return;
+	}
+
+	if (state == METAL_IRQ_ENABLE) {
+		struct sigevent event;
+		SIGEV_PULSE_INIT(&event, self_coid, INTR_PRIORITY, INTR_PULSE_CODE, irq);
+		irqs_ids[irq] = InterruptAttachEvent(irq, &event,_NTO_INTR_FLAGS_NO_UNMASK);
+		if (irqs_ids[irq] == -1) {
+			metal_log(METAL_LOG_ERROR,
+				  "%s: unable to enable irq %d\n", __func__, irq);
+		}
+	} else if (state == METAL_IRQ_DISABLE) {
+		if (irqs_ids[irq] == -1) {
+			metal_log(METAL_LOG_ERROR,
+				  "%s: request to disable irq %d which is not enabled\n",
+				  __func__, irq);
+		} else if (InterruptDetach(irqs_ids[irq]) == -1) {
+			metal_log(METAL_LOG_ERROR,
+				  "%s: unable to disable irq %d\n",
+				  __func__, irq);
+		}
+	} else {
+		metal_log(METAL_LOG_ERROR,
+			  "%s: unknown requested interrupt state %d for irq %d\n",
+			  __func__, state, irq);
+	}
+}
+
+/**
+* @brief       IRQ handler
+* @param[in]   args  not used. required for pthread.
+*/
+static void *metal_qnx_irq_handling(void *args)
+{
+	struct _pulse msg;
+	metal_unused(args);
+
+	while (1) {
+		// block here waiting for interrupt pulse
+		int rcvid = MsgReceivePulse(chid, &msg, sizeof(msg), NULL);
+
+		if (rcvid == -1) {
+			if (!irq_handling_stop) {
+				metal_log(METAL_LOG_ERROR,
+					  "MsgReceivePulse failed: %s\n", strerror(errno));
+			}
+			break;
+		}
+
+		if (msg.code == _PULSE_CODE_DISCONNECT) {
+			ConnectDetach(msg.scoid);
+			continue;
+		}
+
+		if (msg.code != INTR_PULSE_CODE) {
+			metal_log(METAL_LOG_WARNING,
+				  "Unhandled pulse code: %d\n", msg.code);
+			continue;
+		}
+
+		int irq = msg.value.sival_int;
+		int irq_id = irqs_ids[irq];
+
+		if (irq_id != -1) {
+			metal_log(METAL_LOG_DEBUG,
+				  "Received interrupt on vector %d (0x%.4hX)\n",
+				  irq, (uint16_t)irq);
+
+			/* call handler for received irq if enabled */
+			if (metal_irq_handle(&irqs[irq], irq) != METAL_IRQ_HANDLED) {
+				metal_log(METAL_LOG_WARNING,
+					  "IRQ %d (0x%.4hX) unhandled\n",
+					  irq, (uint16_t)irq);
+			}
+
+			/* enable receiving of new interrupts */
+      int ret = __QNX__ < 800 ? InterruptUnmask(irq, irq_id) : InterruptUnmask(0, irq_id);
```

Comment:
`__QNX__` is only used to represent the API version starting from QNX 8.0.
On QNX 7.1, it is evaluated as `1`.

```suggestion
      int ret = __QNX__ >= 800 ? InterruptUnmask(0, irq_id) : InterruptUnmask(irq, irq_id);
```